flang/lib/Lower/OpenMP.cpp

```diff
@@ -1752,11 +1754,82 @@ bool ClauseProcessor::processMap(
                                        semanticsContext, stmtCtx, ompObject,
                                        clauseLocation, asFortran, bounds);
 
+          auto checkIfStructComponent =
+              [](const Fortran::parser::OmpObject &ompObject) {
+                bool isComponent = false;
+                std::visit(
+                    Fortran::common::visitors{
+                        [&](const Fortran::parser::Designator &designator) {
+                          if (auto *structComp = Fortran::parser::Unwrap<
+                                  Fortran::parser::StructureComponent>(
+                                  designator)) {
+                            if (std::holds_alternative<Fortran::parser::Name>(
+                                    structComp->base.u))
+                              isComponent = true;
+                          }
+                        },
+                        [&](const Fortran::parser::Name &name) {}},
+                    ompObject.u);
+
+                return isComponent;
+              };
+
+          // TODO: Currently, it appears there's missing symbol information
+          // and bounds information for allocatables and pointers inside
+          // of derived types. The latter needs some additional support
+          // added to the bounds generation whereas the former appears
+          // that it could be a problem when referring to pointer members
+          // via an OpenMP map clause, for the moment we do not handle
+          // these cases and must emit an error.
+          if (checkIfStructComponent(ompObject) &&
+              Fortran::semantics::IsAllocatableOrPointer(
+                  *getOmpObjectSymbol(ompObject)))
+            TODO(currentLocation,
+                 "pointer members of derived types are currently unmapped");
+
+          if (Fortran::semantics::IsAllocatableOrPointer(
+                  *getOmpObjectSymbol(ompObject))) {
+            // We mimic what will eventually be a structure containing a
+            // pointer mapping for allocatables/pointers/target e.g.:
+            //
+            // !$omp target map(from:in, in%map_ptr)
+            //
+            // ===>
+            //
+            // map_entry varptr(in) ....
+            // map_entry varptr(map_ptr) varptrptr(in) ...
+            //
+            // This is to attempt to keep the lowering of these consistent
+            // with structures containing pointers that are mapped like the
+            // example above, where we break it into the descriptor being the
+            // main "structure" being mapped and the contained pointer the
+            // specific member being referenced. This is of course implicit,
```

Comment:
You note that this is implicit - this should be reflected in the mapping operation too.

Reply:
In what sense do you mean? Currently we're aiming to add implicit captures to the map list as well, not just explicit maps, as a sort of canonicalization process.
 
Do you mean we should not specially handle this case at the PFT -> MLIR lowering level and handle it later on in the MLIR -> LLVM-IR lowering process (or somewhere else)?

Reply:
This point is simply to distinguish between user mapping and compiler implied mapping. Currently in the generated IR, there's no way to distinguish.

Regarding the question about where to handle this - I suggest somewhere after lowering and before LLVM-IR generation (aka at the MLIR level). Lowering IMO should reflect the original user's code (and the user intended to map the data). Also it should not force a decision on how to get descriptor on device - like I noted, we may want to firstprivatize it in some cases.

Reply:
Thank you very much for the clarification, I believe with the current work on IsolatedFromAbove the rest of the implicit maps will be raised into the map entries during the initial generation of the IR, but I think that's currently still under investigation and discussion. 

However, for this particular case, perhaps as you suggested in the main comment on the patch, an MLIR pass is perhaps more apt if the lowered IR must maintain as much similarity to the originally written code as possible. However, the current series of OpenMP passes occurs before the first verification for the lowered IR occurs so I imagine that the emitted IR would still not reflect the users original code if we placed it there (perhaps there's a way to dump the IR earlier though).

Reply:
I have not quite followed the IsolatedFromAbove change, but I understand what the issue would be in this case - namely it is not possible to do a live-in analysis since there can be no outside definitions. In this case, I suggest you update the map_info to ensure you can capture that the descriptor mapping is implicit and still generate it during lowering. Regarding the mapping action to apply, I suggest you add an empty placeholder that can be filled in by a follow-up pass which can decide whether it applies "to", "tofrom", or "firstprivate". Conservatively, you can make it "tofrom" during lowering - even though I don't necessarily prefer this option.

Reply:
Thank you for the input! I think this is roughly what's done currently without the follow up pass, however, I still need too see exactly how the IsolatedFromAbove changes exactly mesh with my current patches (other patch using your awesome bounds work is the array sectioning patch I have up, so do feel free to review that as well if you wish, I didn't add you as I wasn't sure you'd want to/have the time), so I imagine some revisions of them will be required after the patch lands. 

As for the mapping action I'm happy making it "to" initially if this is what you'd prefer and it's possible (especially as I believe the previous example on pointer re-association is quite possibly illegal in OpenMP 5.0-5.2 after a re-reading of the data sharing portion of the specification). I think the firstprivate case may be a little further down the line as I'm not entirely sure how well it's supported in current Fortran+OpenMP